File: Cpp/Legacy89DiskKit.Cpp/src/hu_basic_read_rules.cpp
#include "legacy89diskkit/cpp/hu_basic_read_rules.hpp"

#include "legacy89diskkit/cpp/hu_basic_fat_rules.hpp"

namespace legacy89diskkit::cpp
{
std::vector<std::uint8_t> HuBasicReadRules::TrimToRecordedLength(const std::vector<std::uint8_t>& data, const HuBasicFileEntry& file_entry)
{
    if (file_entry.size > 0 && data.size() > file_entry.size)
    {
        return std::vector<std::uint8_t>(data.begin(), data.begin() + static_cast<std::ptrdiff_t>(file_entry.size));
    }

    return data;
}

std::vector<std::uint8_t> HuBasicReadRules::TrimToTerminalLength(
    const std::vector<std::uint8_t>& data,
    DiskType disk_type,
    const HuBasicConfiguration& config,
    int cluster_count,
    int terminal_flag,
    std::uint32_t recorded_size)
{
    if (disk_type != DiskType::TwoHD)
    {
        return data;
    }

    const auto used_in_last = HuBasicFatRules::GetLastClusterUsedSectors(terminal_flag);
    if (used_in_last == 0)
    {
        return data;
    }

    const auto sectors_per_cluster = config.cluster_size / config.sector_size;
    const auto total_records = (cluster_count - 1) * sectors_per_cluster + used_in_last;
    const auto total_bytes = total_records * config.sector_size;

    if (recorded_size == 0 || total_bytes < static_cast<int>(data.size()))
    {
        return std::vector<std::uint8_t>(data.begin(), data.begin() + total_bytes);
    }

    return data;
}

std::vector<std::uint8_t> HuBasicReadRules::ExtractAsciiPayload(const std::vector<std::uint8_t>& data)
{
    std::vector<std::uint8_t> result;
    result.reserve(data.size());

    for (const auto value : data)
    {
        if (value == 0x1a)
        {
            break;
        }

        result.push_back(value);
    }

    return result;
}

std::vector<std::uint8_t> HuBasicReadRules::ResolveReadPayload(
    const std::vector<std::uint8_t>& data,
    const HuBasicFileEntry& file_entry,
    DiskType disk_type,
    const HuBasicConfiguration& config,
    int cluster_count,
    int terminal_flag)
{
    auto adjusted = TrimToTerminalLength(data, disk_type, config, cluster_count, terminal_flag, file_entry.size);
    if (file_entry.attributes.is_ascii)
    {
        return ExtractAsciiPayload(adjusted);
    }

    return TrimToRecordedLength(adjusted, file_entry);
}
}

File: CSharp/Legacy89DiskKit.FileSystem.Infrastructure/FileSystem/HuBasic/HuBasicReadRules.cs
using Legacy89DiskKit.DiskImage.Domain.Model;
using Legacy89DiskKit.FileSystem.Domain.Model;
using Legacy89DiskKit.FileSystem.Infrastructure.HuBasic.Models;

namespace Legacy89DiskKit.FileSystem.Infrastructure.HuBasic;

public static class HuBasicReadRules
{
    public static byte[] TrimToRecordedLength(byte[] data, FileEntry file)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));
        if (file == null) throw new ArgumentNullException(nameof(file));

        return file.Size > 0 && data.Length > file.Size
            ? data.Take((int)file.Size).ToArray()
            : data;
    }

    public static byte[] TrimToTerminalLength(byte[] data, DiskType diskType, HuBasicConfiguration config, int clusterCount, int terminalFlag, long recordedSize)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));
        if (config == null) throw new ArgumentNullException(nameof(config));

        if (diskType != DiskType.TwoHD)
        {
            return data;
        }

        int usedInLast = HuBasicFatRules.GetLastClusterUsedSectors(terminalFlag);
        if (usedInLast == 0)
        {
            return data;
        }

        int sectorsPerCluster = config.ClusterSize / config.SectorSize;
        int totalRecords = (clusterCount - 1) * sectorsPerCluster + usedInLast;
        int totalBytes = totalRecords * config.SectorSize;

        if (recordedSize == 0 || totalBytes < data.Length)
        {
            return data.Take(totalBytes).ToArray();
        }

        return data;
    }

    public static byte[] ExtractAsciiPayload(byte[] data)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));

        var result = new List<byte>(data.Length);
        foreach (byte b in data)
        {
            if (b == 0x1A)
            {
                break;
            }

            result.Add(b);
        }

        return result.ToArray();
    }

    public static byte[] ResolveReadPayload(byte[] data, FileEntry file, DiskType diskType, HuBasicConfiguration config, int clusterCount, int terminalFlag)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));
        if (file == null) throw new ArgumentNullException(nameof(file));
        if (config == null) throw new ArgumentNullException(nameof(config));

        byte[] adjusted = TrimToTerminalLength(data, diskType, config, clusterCount, terminalFlag, file.Size);

        if (file.Attributes.IsAscii)
        {
            return ExtractAsciiPayload(adjusted);
        }

        return TrimToRecordedLength(adjusted, file);
    }
}
